pkg/assets/builder.go

```diff
@@ -21,43 +21,62 @@ import (
 	"fmt"
 	"net/url"
 	"os"
+	"path"
 	"strings"
 
+	"github.com/golang/glog"
 	"k8s.io/kops/pkg/apis/kops"
 	"k8s.io/kops/pkg/featureflag"
 	"k8s.io/kops/pkg/kubemanifest"
+	"k8s.io/kops/util/pkg/hashing"
+	"k8s.io/kops/util/pkg/vfs"
 )
 
 // RewriteManifests controls whether we rewrite manifests
 // Because manifest rewriting converts everything to and from YAML, we normalize everything by doing so
 var RewriteManifests = featureflag.New("RewriteManifests", featureflag.Bool(true))
 
-// AssetBuilder discovers and remaps assets
+// AssetBuilder discovers and remaps assets.
 type AssetBuilder struct {
 	ContainerAssets []*ContainerAsset
 	FileAssets      []*FileAsset
 	AssetsLocation  *kops.Assets
+	// we'd like to use cloudup.Phase here, but that introduces a go cyclic dependency
+	Phase string
 }
 
+// ContainerAsset models a container's location.
 type ContainerAsset struct {
 	// DockerImage will be the name of the docker image we should run, if this is a docker image
 	DockerImage string
 
-	// CanonicalLocation will be the source location of the image, if we should copy it to the actual location
+	// CanonicalFileURLLocation will be the source location of the image, if we should copy it to the actual location
 	CanonicalLocation string
 }
 
+// FileAsset models a file's location.
 type FileAsset struct {
-	// File will be the name of the file we should use
-	File string
+	// FileURL is the URL of a file.
+	FileURL *url.URL
 
```

Comment:
I renamed this struct members as you requested in a different PR @justinsb

Reply:
What was the PR? I think I had other comments in that PR as well..

Reply:
I have made the updates in this PR, apologize for the multiple PRs, but I was trying to break up the work.

Reply:
I will put it here in a but

Reply:
https://github.com/kubernetes/kops/pull/3670